packages/core/src/chopsticks-provider.ts

```diff
@@ -1,286 +1,245 @@
-//           } catch (error) {
-//             handler.callback(error as Error, undefined)
-//           }
-//           delete this.#handlers[e.data.id]
-//         }
-//         break
-
-//       default:
-//         break
-//     }
-//   }
-// }
+import { EventEmitter } from 'eventemitter3'
+import {
+  ProviderInterface,
+  ProviderInterfaceCallback,
+  ProviderInterfaceEmitCb,
+  ProviderInterfaceEmitted,
+  ProviderStats,
+} from '@polkadot/rpc-provider/types'
+
+import { Blockchain } from './blockchain'
+import { StorageValues, setStorage } from './utils'
+import { allHandlers } from './rpc'
+import { defaultLogger } from './logger'
+import { setup } from './setup'
+
+const providerHandlers = {
+  ...allHandlers,
+  new_block: async (context: any, _params: any, _subscriptionManager: any) => {
+    const { chain } = context
+    const block = await chain.newBlock()
+    return block
+  },
+}
+
+const logger = defaultLogger.child({ name: '[Chopsticks provider]' })
+
+interface SubscriptionHandler {
+  callback: ProviderInterfaceCallback
+  type: string
+}
+
+interface Subscription extends SubscriptionHandler {
+  method: string
+  params?: unknown[]
+  onCancel?: () => void
+  result?: unknown
+}
+
+export interface ChopsticksProviderProps {
+  /** upstream endpoint */
+  endpoint: string
+  /** default to latest block */
+  blockHash?: string
+  dbPath?: string
+  storageValues?: StorageValues
+}
+
+/**
+ * A provider for ApiPromise.
+ *
+ * Currectly only support browser environment.
+ */
+export class ChopsticksProvider implements ProviderInterface {
+  #isConnected = false
+  #eventemitter: EventEmitter
+  #isReadyPromise: Promise<void>
+  #endpoint: string
+  readonly stats?: ProviderStats
+  #subscriptions: Record<string, Subscription> = {}
+  #blockHash: string | undefined
+  #dbPath: string | undefined
+  #storageValues: StorageValues | undefined
+  #chain: Blockchain | undefined
+
+  constructor({ endpoint, blockHash, dbPath, storageValues }: ChopsticksProviderProps) {
+    if (!endpoint) {
+      throw new Error('ChopsticksProvider requires the upstream endpoint')
+    }
+    this.#endpoint = endpoint
+    this.#blockHash = blockHash
+    this.#dbPath = dbPath
+    this.#storageValues = storageValues
+
+    this.#eventemitter = new EventEmitter()
+
+    this.#isReadyPromise = new Promise((resolve, reject): void => {
+      this.#eventemitter.once('connected', (): void => {
+        logger.debug('isReadyPromise: connected.')
+        resolve()
+      })
+      this.#eventemitter.once('error', reject)
+    })
+
+    this.connect()
+  }
+
+  get hasSubscriptions(): boolean {
+    return true
+  }
+
+  get isClonable(): boolean {
+    return true
+  }
+
+  get isConnected(): boolean {
+    return this.#isConnected
+  }
+
+  get isReady(): Promise<void> {
+    return this.#isReadyPromise
+  }
+
+  clone = (): ProviderInterface => {
+    return new ChopsticksProvider({ endpoint: this.#endpoint })
```

Comment:
This doesn't feel right. Cloning provider will duplicate the chain. Maybe we can make provider to accept either endpoint or chainPromise so when cloning we create a provider with chainPromise